canvas/biplot.py

```diff
@@ -0,0 +1,499 @@
+        if feature_color_category is None:
+            feature_color_category = 'Taxonomy'
+
+        for name, group in
+        feature_df.groupby(feature_df_metadata[feature_color_category]):
+            if len(name.split(';')) > 1:
+                otu_feature = name.split(';')[taxa_rank].replace(' ', '')
+                otu_feature_append.append(otu_feature)
+            else:
+                otu_feature = name
+                otu_feature_append.append(otu_feature)
+
+        otu_feature_append = sorted(list(set(otu_feature_append)))
+        cNorm = colors.Normalize(vmin=0, vmax=(len(otu_feature_append)-1))
+        scalarMap = cmx.ScalarMappable(norm=cNorm, cmap=feature_colors)
+
+        for index, row in enumerate(otu_feature_append):
+            colorVal.append(scalarMap.to_rgba(index))
+        feature_color_dict = dict(zip(otu_feature_append, colorVal))
+        feature_color_dict = OrderedDict(sorted(feature_color_dict.items(),
+                                                key=lambda x: x[0]))
+
+        otu_feature_append = []
+        for name, group in
+        feature_df.groupby(feature_df_metadata[feature_color_category]):
+            if len(name.split(';')) > 1:
+                otu_feature = name.split(';')[taxa_rank].replace(' ', '')
+                otu_feature_append.append(otu_feature)
+                ax2.arrow(0, 0,
+                          np.asscalar(group[feature_df_x]),
+                          np.asscalar(group[feature_df_y]),
+                          width=arrow_width, head_width=arrow_head,
+                          alpha=feat_alpha,
+                          color=feature_color_dict[otu_feature])
+            else:
+                otu_feature = name
+                otu_feature_append.append(otu_feature)
+                ax2.arrow(0, 0,
+                          np.asscalar(group[feature_df_x]),
+                          np.asscalar(group[feature_df_x]),
+                          width=arrow_width, head_width=arrow_head,
+                          alpha=feat_alpha,
+                          color=feature_color_dict[otu_feature])
+
+        for key, value in feature_color_dict.items():
+            recs.append(mpatches.Rectangle((0, 0), 1, 1,
+                                           fc=feature_color_dict[key],
+                                           alpha=feat_alpha))
+            feature.append(key)
+        ax2.legend(recs, feature, loc=feat_leg_loc,
+                   title=feature_color_category)
+        feature_color_dict = None
+
+    if feature_df_metadata is not None and
+    feature_color_dict is not None and
+    feature_color_category is not None:
+        print('Feature Block 3')
+        for name, group in
+        feature_df.groupby(feature_df_metadata[feature_color_category]):
+            if len(name.split(';')) > 1:
+                otu_feature = name.split(';')[taxa_rank].replace(' ', '')
+                otu_feature_append.append(otu_feature)
```

Comment:
Umm, this looks extremely repetitive.  Let's discuss tomorrow how we can start cleaning this up.